dirty_cat/fuzzy_join.py

```diff
@@ -0,0 +1,109 @@
+# TODO: add HashingVectorizer as an option.
+
+"""
+Fuzzy joining tables using string columns.
+The principle is as follows:
+  1. We transform the key columns strings to vectors using CountVectorizer.
+  2. For each category, we use the nearest neighbor method to find its closest
+  neighbor and establish a match.
+  3. We match the tables using the previous information.
+Categories from the two tables that share many n-grams have greater
+probability of beeing matched together. The join is based on
+morphological similarities between strings.
+"""
+
+import pandas as pd
+import numpy as np
+from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
+from sklearn.neighbors import NearestNeighbors
+from sklearn.base import BaseEstimator, TransformerMixin
+
+
+class FuzzyJoin(BaseEstimator, TransformerMixin):
+    """
+    Join tables based on categorical string columns as joining keys.
+
+    Parameters
+    ----------
+
+    analyzer : str, default='char'.
+        Analyzer parameter for the CountVectorizer.
+        Options: {‘word’, ‘char’, ‘char_wb’}, describing whether the matrix V
+        to factorize should be made of word counts or character n-gram counts.
+        Option ‘char_wb’ creates character n-grams only from text inside word
+        boundaries; n-grams at the edges of words are padded with space.
+    ngram_range : tuple (min_n, max_n), default=(2, 4)
+        The lower and upper boundary of the range of n-values for different
+        n-grams to be extracted. All values of n such that min_n <= n <= max_n.
+        will be used.
+
+    """
+
+    def __init__(self, analyzer="char", ngram_range=(2, 4)):
+        self.ngram_range = ngram_range
+        self.analyzer = analyzer
+
+    def join(self, left_table, right_table, on, return_distance=False):
+        """Join left and right table.
+
+        Parameters
+        ----------
+        left_table: pd.DataFrame
+            Reference table on which the join will be performed.
```

Comment:
I think in our case it doesn't need to be a reference table (unlike in autofj where they assume that the left table has few duplicates).

Reply:
I see, you are right, I will remove the term reference.